teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,475 @@
+
+## Alternatives
+
+### 1. Implicitly Project `Parameters` as Environment Variables in `Steps`
+
+We have projected `Parameters` as environment variables in the `git-clone` Task in the
+Catalog and in our own plumbing repository. We can use this approach, and simplify the
+syntax by implicitly projecting `Parameters` as environment variables in `Steps`.
+The environment variable should be prefixed with `PARAM_` to make it clear and predictable.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo ${PARAM_BAR}
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}      
+```
+
+##### Discussion
+
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].
+
+### 2. Implicitly Project `Parameters` into the Filesystem of `Steps`
+
+Write `Parameters` in files - `/tekton/parameters/in` - mounted into `Steps`
+in the same way that `Results` are supplied. However, instead of exposing the
+file path, we would hide it behind a variable (possibly reusing `$(params.<name>)`.
+
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+```
+
+This approach is also under discussion in the context of changing the way `Result`
+values are stored in [TEP-0086][tep-0086-alt]. In that proposal a sidecar is
+responsible for fetching `Results` from previous `Tasks` and writing them to
+a location on disk for use as `Parameters`.
+
+##### Discussion
+
+This approach aligns well with the proposal in [TEP-0086][tep-0086-prop], and
+brings consistency in how we supply both `Parameters` and `Results` to `Steps`.
+
+### 3. Provide an Escaped Form of `Parameters` as Tekton Variables for Injection
+
+This was the proposal in [TEP-0017][tep-0017].
+
+> In a Task or Pipeline, for each parameter `foo`, provide `$(params.foo.shell-escaped)`.
+>
+> This variable shall be escaped according to `printf '%q'` rules as used in bash.
+> For example, the literal string `Hello '\!"` becomes `Hello\ \'\\\!\"`. A script
+> which ran `echo $(params.foo.shell-escaped)` would print the literal string. The
+> escaped value would do the wrong thing when placed inside double or single quotes.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar.shell-escaped)
+```
+
+##### Discussion
+
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].
```

Comment:
Will shell escaping prevent all shell injection vulnerabilities? Our [catalog guidance](https://github.com/tektoncd/catalog/blob/main/recommendations.md#dont-use-interpolation-in-scripts-or-string-arguments) says that "no amount of escaping will be air-tight".

This sounds safer than current behavior but it seems like it will still lead to users having to audit Tasks for security.